ambry-server/src/main/java/com.github.ambry.server/AmbryServerRequests.java

```diff
@@ -145,7 +155,27 @@ public void handleRequests(Request request) throws InterruptedException {
    */
   private boolean isRequestEnabled(RequestOrResponseType requestType, PartitionId id) {
     Set<PartitionId> requestDisableInfo = requestsDisableInfo.get(requestType);
-    return requestDisableInfo == null || !requestDisableInfo.contains(id);
+    // 1. check if request is disabled by admin request
+    if (requestDisableInfo != null && requestDisableInfo.contains(id)) {
+      return false;
+    }
+    if (serverConfig.serverValidateRequestBasedOnStoreState) {
+      // 2. check if request is disabled due to current state of store
+      Store store = storeManager.getStore(id);
+      if (requestType == RequestOrResponseType.PutRequest && !POST_ALLOWED_STORE_STATES.contains(
+          store.getCurrentState())) {
+        logger.error("{} is not allowed because current state of store {} is {}", requestType, id,
```

Comment:
Is error the correct log level for this? How often will this occur during normal operation?

Reply:
Should be infrequent. It happens in two cases:
1. remove replica: when local replica's state transition has completed but frontends still have old view and still issue PUT requests to the store. Once Helix state change notification arrives at frontend, this should die out quickly. 
2. shutdown server: when server is shut down, local replica is directly set to OFFLINE and some frontends may still attempt to write to this store due to Helix notification delay.